kitsune/sumo/i18n.py

```diff
@@ -0,0 +1,135 @@
+    """
+    if not settings.USE_I18N:
+        return list(urls)
+    return [
+        django.urls.URLResolver(
+            LocalePrefixPattern(prefix_default_language=prefix_default_language),
+            list(urls),
+        )
+    ]
+
+
+def split_into_language_and_path(path):
+    """
+    Given a URL path that starts with a language, returns the
+    language and the rest of the path after the language.
+    """
+    language, slash, rest_of_path = path.lstrip("/").partition("/")
+    return (language, f"{slash}{rest_of_path}")
+
+
+def normalize_language(language):
+    """
+    Given a language code, returns the language code supported by SUMO in the
+    proper case, for example "eN-us" --> "en-US" or "sC" --> "it", or None if
+    SUMO doesn't support the language code.
+    """
+    if not language:
+        return None
+    lc_language = language.lower()
+    return settings.LANGUAGE_URL_MAP.get(lc_language) or FALLBACK_LANGUAGE_URL_MAP.get(lc_language)
+
+
+def normalize_path(path, force_language=None):
+    """
+    Normalizes the language code, if any, that starts the path. If "force_language" is
+    given, it will replace any existing language code in the path, or prefix the path
+    if the path didn't already start with a language code.
+
+    Example when "force_language=None":
+        Requested  --> Normalized
+            /en-us --> /en-US  (normalization of case)
+            /fr-ca --> /fr     (supported variant)
+            /en-gb --> /en-US  (supported variant)
+            /sc    --> /it     (explicit fallback)
+            /ak    --> /en-US  (explicit fallback)
+
+    Example when "force_language=de":
+        Requested    --> Normalized
+            /path    --> /de/path  (path prefixed)
+            /it/path --> /de/path  (language code replaced)
+    """
+    # Use Django's standard way to check if the path starts with a supported language.
+    language_from_path = translation.get_language_from_path(path)
+
+    # Even if Django found that the path didn't start with a supported language, the path
+    # may start with one for which SUMO defines a supported fallback language, for example
+    # "sc" --> "it".
+    if not language_from_path and (mo := LANGUAGES_WITH_FALLBACKS_REGEX.match(path)):
+        language_from_path = mo.group("language")
+
+    normalized_language_from_path = normalize_language(language_from_path)
+
+    if normalized_language_from_path:
+        actual_language_from_path, remaining_path = split_into_language_and_path(path)
+        if force_language:
+            return f"/{force_language}{remaining_path}"
+        elif actual_language_from_path != normalized_language_from_path:
+            return f"/{normalized_language_from_path}{remaining_path}"
+    elif force_language:
+        return f"/{force_language}{path}"
+
+    return path
+
+
+def get_language_from_request(request, check_path=False, check_user=False):
```

Comment:
If I remember correctly, I used 3 separate calls to `normalize_language` because it also will return `None` if the language passed-in is not supported, so it's part of the check at each step. I'll look into it again.

Reply:
I refactored it.